velox/exec/Spill.h

```diff
@@ -0,0 +1,340 @@
+  std::unique_ptr<ReadFile> input_;
+  uint64_t size_;
+  BufferPtr buffer_;
+  uint64_t offset_ = 0;
+};
+
+// A source of VectorRows coming either from a file or memory.
+class SpillStream {
+ public:
+  SpillStream(TypePtr type, memory::MemoryPool& pool)
+      : type_(std::dynamic_pointer_cast<const RowType>(type)), pool_(pool) {
+    ordinal_ = ++ordinalCounter_;
+  }
+
+  virtual ~SpillStream() = default;
+
+  virtual bool atEnd() const = 0;
+
+  virtual std::string label() const {
+    return fmt::format("{}", ordinal_);
+  }
+
+  VectorRow next() {
+    if (index_ == numRows_) {
+      nextBatch();
+      index_ = 0;
+      numRows_ = rowVector_->size();
+      SelectivityVector allRows(numRows_);
+      auto width = rowVector_->childrenSize();
+      if (decoded_.size() < width) {
+        decoded_.resize(width);
+      }
+      for (auto i = 0; i < width; ++i) {
+        if (!decoded_[i]) {
+          decoded_[i] = std::make_unique<DecodedVector>();
+        }
+        decoded_[i]->decode(*rowVector_->childAt(i), allRows);
+      }
+    }
+    return {&decoded_, rowVector_.get(), index_++};
+  }
+
+  virtual uint64_t size() const {
+    return 0;
+  }
+
+ protected:
+  virtual void nextBatch() = 0;
+
+  std::shared_ptr<const RowType> type_;
+  memory::MemoryPool& pool_;
+  RowVectorPtr rowVector_;
+  vector_size_t index_ = 0;
+  vector_size_t numRows_ = 0;
+  std::vector<std::unique_ptr<DecodedVector>> decoded_;
+  int32_t ordinal_;
+  static std::atomic<int32_t> ordinalCounter_;
+};
+
+// Represents a spill file that is first in write mode and then
+// turns into an source of VectorRow. Owns a file system file that
+// contains the spilled data and is live for the duration of 'this'.
+class SpillFile : public SpillStream {
+ public:
+  SpillFile(TypePtr type, const std::string& path, memory::MemoryPool& pool)
+      : SpillStream(type, pool), path_(path) {}
+
+  ~SpillFile();
+
+  // Returns a file for writing spilled data.
+  WriteFile& output();
+
+  // Finishes writing and flushes any unwritten data.
+  void finishWrite() {
+    VELOX_CHECK(output_);
+    size_ = output_->size();
+    output_ = nullptr;
+  }
+
+  // Prepares 'this' for reading. Positions the read at the first row of
+  // content.
+  void startRead();
+
+  // Returns the file size in bytes.
+  uint64_t size() const override {
+    if (output_) {
+      return output_->size();
+    }
+    return size_;
+  }
+
+  // True if no more VectorRows can be read from 'this'.
+  bool atEnd() const override {
+    return index_ >= numRows_ && stream_->atEnd();
+  }
+
+  // Sets 'result' to refer to the next row of content of 'this'.
+  void read(RowVector& result);
+
+ protected:
+  void nextBatch() override {
+    VectorStreamGroup::read(stream_.get(), &pool_, type_, &rowVector_);
+    index_ = 0;
+    numRows_ = rowVector_->size();
+  }
+
+ private:
+  std::string path_;
+  uint64_t size_ = 0;
+  std::unique_ptr<WriteFile> output_;
+  std::unique_ptr<SpillInput> stream_;
```

Comment:
stream_ name is not very descriptive as it can be refer to both output and input streams; here it seems like it refers to input stream, hence, might be cleaner as input_ (to match output_).